apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -1006,15 +1074,13 @@ enqueue_transient(ClientInfo, Msg0, Session = #{s := S, props := #{upgrade_qos :
     %% queued messages. Since streams in this DB are exclusive to the
     %% session, messages from the queue can be dropped as soon as they
     %% are acked.
-    Subs = emqx_persistent_session_ds_state:get_subscriptions(S),
-    Msgs = [
-        Msg
-     || SubMatch <- emqx_topic_gbt:matches(Msg0#message.topic, Subs, []),
-        Msg <- begin
-            #{props := SubOpts} = emqx_topic_gbt:get_record(SubMatch, Subs),
-            emqx_session:enrich_message(ClientInfo, Msg0, SubOpts, UpgradeQoS)
-        end
-    ],
+    case emqx_persistent_session_ds_state:get_subscription(Msg0#message.topic, S) of
```

Comment:
Matching topic of the message against subscriptions literally is wrong, but that's what mem session is doing, apparently.

Reply:
I think it's not the same actually. In-mem sessions receive `deliver()`s and use the [_routing_ topic to look up the subscription](https://github.com/emqx/emqx/blob/0f426e6e77273bc18640698defaa0edc856b1cdf/apps/emqx/src/emqx_session.erl#L428) and not the _message_ topic. Those are different, _routing_ topic comes from the routing table and is basically topic filter of some subscription.

Reply:
Ok, I think this will work exactly the same then.

Reply:
This is part of the shared channel logic, isn't it?

Reply:
Oh right, I see now, sorry for the confusion. That actually means that those messages have already been passed through `enrich_message/4` before hitting this function, so this second "enrichment" is not really needed?

Reply:
Thanks, fixed.